File: java/patterns-use-cases/async-tasks-parallelize-work/src/main/java/my/example/parallelizework/FanOutWorker.java
package my.example.parallelizework;

import com.fasterxml.jackson.core.type.TypeReference;
import dev.restate.sdk.Awaitable;
import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;
import dev.restate.sdk.serde.jackson.JacksonSerdes;
import java.util.ArrayList;
import java.util.List;

import static my.example.parallelizework.utils.DataProcessingUtils.aggregate;
import static my.example.parallelizework.utils.DataProcessingUtils.split;

/*
 * Restate makes it easy to parallelize async work by fanning out tasks.
 * Afterwards, you can collect the result by fanning in the partial results.
 * Durable Execution ensures that the fan-out and fan-in steps happen reliably exactly once.
 */
@Service
public class FanOutWorker {

  public record Task(){}
  public record SubTask(){}
  public record SubTaskResult(){}
  public record Result(){}

  @Handler
  public Result run(Context ctx, Task task) {
    // Split the task in subtasks
    List<SubTask> subTasks = ctx.run(JacksonSerdes.of(new TypeReference<>() {}),
            () -> split(task));

    // Fan out the subtasks - run them in parallel
    List<Awaitable<?>> resultFutures = new ArrayList<>();
    for (SubTask subTask : subTasks) {
      resultFutures.add(FanOutWorkerClient.fromContext(ctx).runSubtask(subTask));
    }

    Awaitable.all(resultFutures).await();

    // Fan in - Aggregate the results
    var results = (SubTaskResult[]) resultFutures.stream().map(Awaitable::await).toArray();
    return aggregate(results);
  }

  // Can also run on FaaS
  @Handler
  public SubTaskResult runSubtask(Context ctx, SubTask subTask) {
    // Processing logic goes here ...
    // Can be moved to a separate service to scale independently
    return new SubTaskResult();
  }

  public static void main(String[] args) {
     RestateHttpEndpointBuilder.builder().bind(new FanOutWorker()).buildAndListen(9082);
  }
}

File: go/patterns-use-cases/src/parallelizework/fanoutworker.go
package main

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log/slog"
	"os"
)

/**
 * Restate makes it easy to parallelize async work by fanning out tasks.
 * Afterward, you can collect the result by fanning in the partial results.
 *          +------------+
 *          | Split task |
 *          +------------+
 *                |
 *        ---------------------------------
 *        |                |              |
 * +--------------+ +--------------+ +--------------+
 * | Exec subtask | | Exec subtask | | Exec subtask |
 * +--------------+ +--------------+ +--------------+
 *        |                |               |
 *        ---------------------------------
 *                |
 *          +------------+
 *          | Aggregate  |
 *          +------------+
 * Durable Execution ensures that the fan-out and fan-in steps happen reliably exactly once.
 */

type FanOutWorker struct{}

func (FanOutWorker) Run(ctx restate.Context, task Task) (Result, error) {
	// Split the task into subtasks
	subtasks, err := split(task)
	if err != nil {
		return Result{}, err
	}

	// Fan out the subtasks - run them in parallel
	subtaskFutures := make([]restate.Selectable, 0, len(subtasks))
	for _, subtask := range subtasks {
		subtaskFutures = append(subtaskFutures,
			restate.Service[SubTaskResult](ctx, "FanOutWorker", "RunSubtask").RequestFuture(subtask))
	}

	selector := restate.Select(ctx, subtaskFutures...)

	// Fan in - Aggregate the results
	subResults := make([]SubTaskResult, 0, len(subtasks))
	for selector.Remaining() {
		response, err := selector.Select().(restate.ResponseFuture[SubTaskResult]).Response()
		if err != nil {
			return Result{}, err
		}
		subResults = append(subResults, response)
	}

	// Fan in - Aggregate the results
	return aggregate(subResults)
}

// RunSubtask can also run on FaaS
func (FanOutWorker) RunSubtask(ctx restate.Context, subtask SubTask) (SubTaskResult, error) {
	// Processing logic goes here ...
	// Can be moved to a separate service to scale independently
	return executeSubtask(ctx, subtask)
}

func main() {
	server := server.NewRestate().
		Bind(restate.Reflect(FanOutWorker{}))

	if err := server.Start(context.Background(), ":9080"); err != nil {
		slog.Error("application exited unexpectedly", "err", err.Error())
		os.Exit(1)
	}
}
